draft-ietf-masque-connect-udp-listen.md

```diff
@@ -153,11 +163,114 @@ UDP Payload:
 : The unmodified UDP Payload of this proxied UDP packet (referred to as "data
 octets" in {{UDP}}).
 
+# Context ID {#contextid}
+
+A Context ID of value 0 represents no compression. When using context ID 0, the
+client and server must specify the IP version, IP and port of the target per
+datagram, see {{format}}. The client can compress IP and port information using
+non zero Context IDs as defined under {{compression}}.
+Receiving and forwarding from Context ID 0 is enabled by default, but may be
+disabled using {{restrictingips}}.
+
+## Header Compression {#compression}
+
+The client MAY choose to map IP and port information per datagram against the
+Context ID defined in {{contextid}}. In such a case, the client MUST send a
+COMPRESSION_REQUEST capsule (see {{capsulerequestformat}}) with the target
+information (see {{targetmappingformat}}) it wishes to compress and the proxy
+MUST respond with either a COMPRESSION_ASSIGN (see {{capsuleresponseformat}})
+capsule if it accepts the compression request, or a COMPRESSION_REJECT
+(see {{capsulerejectformat}}) if it doesn't wish to support compression
+(For example, due to considerable memory requirements of establishing a list
+of mappings per target per client). If the compression was rejected,
+the client and server MUST continue to use uncompressed, Context ID 0 Datagrams
+for the requested target.
+Otherwise, they must exchange compressed datagrams for the given datagram as
+described in {{mappings}}.
+
+### Compression Mapping {#mappings}
+
+If a COMPRESSION_ASSIGN is received by client, the proxy and client MUST
+maintain a compression mapping from Target to Context ID and update this
+mapping whenever new COMPRESSION_REQUEST and COMPRESSION_ASSIGN capsules
+are exchanged between the two. Upon completing this exchange, the proxy and
+the client MUST send each other packets corresponding to a target using the
+unique Context ID assigned to the given target, while omitting IP and port
+information.
+
+
+## Restricting IPs {#restrictingips}
+
+The client MAY request the proxy reject any uncompressed targets by sending
+it the COMPRESSION_CLOSE (see {{compressioncloseformat}}) capsule with the
+enable field set to true. Upon receiving the request, the proxy MUST echo
+back the COMPRESSION_CLOSE with the enable field set to true. Hereon, the
+proxy MUST not forward any packets from uncommpressed targets to the client
+and MUST drop any UDP  payloads received from unregistered targets.
+
+Likewise, the client MAY choose to re-enable traffic from uncompressed
+targets, by sending the COMPRESSION_CLOSE capsule with the enable field
+set to false and the proxy MUST respond with COMPRESSION_CLOSE with enable
+set to false and MUST re-start forwarding uncompressed context ID 0 payloads.
+
+
+## Capsules {#capsules}
+The Listener capsule types are defined below:
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_REQUEST (0x04),
```

Comment:
I'm not sure I understand why we need four capsule types. Could we make this work with only two (COMPRESSION_OPEN and COMPRESSION_CLOSE)? Conceptually you need an open one (sent in one direction to request, and echoed to confirm) and a close one (used as a response to an open to reject, and used directly to request closure)

Reply:
Yes, following this approach now